lib/src/cache/cache.dart

```diff
@@ -0,0 +1,245 @@
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+
+import 'package:w_common/disposable.dart';
+import 'package:w_common/func.dart';
+
+/// Immutable payload that indicates a change in a [Cache].
+class CacheContext<TIdentifier, TValue> {
+  /// Identifies a single instance of a [TValue]
+  final TIdentifier id;
+
+  /// The current value stored in the [Cache] for the attributed [TIdentifier].
+  final TValue value;
+
+  CacheContext(this.id, this.value);
+}
+
+/// Maintians a reference to a given any [TValue] attributed to a [TIdentifier].
+///
+/// References are retained for the lifecycle of the instance of the [Cache],
+/// unless explicitly removed.
+class Cache<TIdentifier, TValue> extends Object with Disposable {
+  /// The backing store for values in the [Cache].
+  Map<TIdentifier, Future<TValue>> _cache = <TIdentifier, Future<TValue>>{};
+
+  // ignore: close_sinks
+  StreamController<CacheContext<TIdentifier, TValue>> _didUpdateController =
+      new StreamController<CacheContext<TIdentifier, TValue>>.broadcast();
+
+  // ignore: close_sinks
+  StreamController<CacheContext<TIdentifier, TValue>> _didRemoveController =
+      new StreamController<CacheContext<TIdentifier, TValue>>.broadcast();
+
+  /// A collection of pending removals for a given [TIdentifier].
+  Map<TIdentifier, Completer<Null>> _removalCompleters =
+      <TIdentifier, Completer<Null>>{};
+
+  Cache() {
+    [_didRemoveController, _didUpdateController]
+        .forEach(manageStreamController);
+  }
+
+  /// The collection of removals from the [Cache].
+  Stream<CacheContext<TIdentifier, TValue>> get didRemove =>
+      _didRemoveController.stream;
+
+  /// The collection of changes to the [Cache].
+  ///
+  /// Values are emitted when the cached value is changes.
+  Stream<CacheContext<TIdentifier, TValue>> get didUpdate =>
+      _didUpdateController.stream;
+
+  /// Returns a value from the cache for a given [TIdentifier] or adds the
+  /// [TValue] returned by the value factory when the value does not exist in
+  /// the cache.
```

Comment:
It adds *and* returns the value from the value factory if the identifier wasn't found, right? Maybe clarify this.